active_discussion/embedded/static-unsynchronized.md

```diff
@@ -0,0 +1,82 @@
+# Unsynchronized access to a shared `static mut` variable
+
+Sometimes interrupt handlers need to share state. This is done using a static
+variable. We claim that *no* synchronization is needed to access the shared
+static variable if the interrupt handlers run at the same priority.
+
+Note that:
+
+- The programs in this document target the [Basic Embedded Abstract Machine
+  (BEAM)][beam]. Please become familiar with the linked specification before you
+  read the rest of this document.
+
+[beam]: https://github.com/rust-lang/unsafe-code-guidelines/pull/111
+
+- In these programs we assume that [rust-lang/rfcs#2585][rfc2585] has been
+  accepted and implemented.
+
+[rfc2585]: https://github.com/rust-lang/rfcs/pull/2585
+
+## Example program
+
+Consider this program
+
+``` rust
+#![no_std]
+
+#[no_mangle]
+unsafe fn main() -> ! {
+    unsafe {
+        asm!("ENABLE_INTERRUPTS" : : : : "volatile");
+    }
+
+    loop {
+        // .. any safe code ..
+    }
+}
+
+static mut X: u128 = 0;
+
+#[no_mangle]
+unsafe fn INTERRUPT0() {
+    let x: &mut u128 = unsafe { &mut X };
+
+    // .. any safe code ..
+}
+
+#[no_mangle]
+unsafe fn INTERRUPT1() {
+    let x: &mut u128 = unsafe { &mut X };
+
+    // .. any safe code ..
+}
+```
+
+Note that "any safe code" can *not* call `main`, `INTERRUPT0` or `INTERRUPT1`
+(because they are `unsafe` functions), use `asm!` or access registers.
+
+**Claim**: this program is well-defined / sound. Rust aliasing rules are
+preserved because no preemption is possible between `INTERRUPT0` and
+`INTERRUPT1`.
```

Comment:
IDK exactly what you mean by "no preemption is possible between `INTERRUPT0` and
`INTERRUPT1`.", but from the BEAM document, if `INTERRUPT1` has higher priority than `INTERRUPT0`, it can preempt `INTERRUPT0`.

Reply:
Changing the priority is an `unsafe` operation. The invariant required to keep this safe / sound is that interrupt priorities won't be changed.

I didn't paint the whole picture here, sorry about that. In practice, priorities are exposed as something that the user sets at compile time and then never changes at runtime. Before the user code runs some runtime / framework configures the interrupt priorities before globally enabling interrupts (both operations are `unsafe`). The user code is not meant to change interrupt priorities; the API to do that is `unsafe` because it can cause UB if it enables preemption where previously there wasn't.